pkg/util/deeptest/statictesthelper.go

```diff
@@ -0,0 +1,181 @@
+	if h.shouldIgnore(path) {
+		return
+	}
+
+	// This function assumes that `a` and `b` are the same type
+	switch valA.Type().Kind() {
+	case reflect.Struct:
+		for i := 0; i < valA.NumField(); i++ {
+			h.assertEqual(t, valA.Field(i), valB.Field(i), path+"."+valA.Type().Field(i).Name)
+		}
+	case reflect.Ptr:
+		// both of them are not nil
+		require.NotEqual(t, 0, valA.Pointer(), path+" should not be nil")
+		require.NotEqual(t, 0, valB.Pointer(), path+" should not be nil")
+		// pointer type should be different
+		require.NotSame(t, valA.Pointer(), valB.Pointer(), path+" should be different")
+
+		h.assertEqual(t, valA.Elem(), valB.Elem(), path)
+	case reflect.Slice, reflect.Array:
+		require.Equal(t, valA.Len(), valB.Len(), path+" should have the same length")
+		for i := 0; i < valA.Len(); i++ {
+			h.assertEqual(t, valA.Index(i), valB.Index(i), path+fmt.Sprintf("[%d]", i))
+		}
+	case reflect.Bool:
+		require.Equal(t, valA.Bool(), valB.Bool(), path+" should be the same")
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		require.Equal(t, valA.Int(), valB.Int(), path+" should be the same")
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		require.Equal(t, valA.Uint(), valB.Uint(), path+" should be the same")
+	case reflect.Float32, reflect.Float64:
+		require.Equal(t, valA.Float(), valB.Float(), path+" should be the same")
+	case reflect.String:
+		require.Equal(t, valA.String(), valB.String(), path+" should be the same")
+	case reflect.Map:
+		require.Equal(t, valA.Len(), valB.Len(), path+" should have the same length")
+		for _, key := range valA.MapKeys() {
+			h.assertEqual(t, valA.MapIndex(key), valB.MapIndex(key), path+fmt.Sprintf("[%v]", key))
+		}
+	case reflect.Interface:
+		h.assertEqual(t, valA.Elem(), valB.Elem(), path)
+	case reflect.Func:
+		require.NotEqual(t, 0, valA.Pointer(), path+" should not be nil")
+		require.NotEqual(t, 0, valB.Pointer(), path+" should not be nil")
+		require.NotSame(t, valA.Pointer(), valB.Pointer(), path+" should be different")
+		return
+	default:
+		require.Fail(t, "unsupported type", path)
+	}
+}
+
+// Option is the option for the deep test.
+type Option interface {
+	apply(*staticTestHelper)
+}
+
+type ignorePathOption struct {
+	ignorePath []string
+}
+
+func (o ignorePathOption) apply(h *staticTestHelper) {
+	h.ignorePath = make([]regexp.Regexp, 0, len(o.ignorePath))
+	for _, path := range o.ignorePath {
+		h.ignorePath = append(h.ignorePath, *regexp.MustCompile(path))
+	}
+}
+
+// WithIgnorePath specifies the paths that should be ignored during the deep test.
+func WithIgnorePath(ignorePath []string) Option {
+	return ignorePathOption{ignorePath}
+}
+
+// AssertEveryFieldNotEqual asserts that every field of `a` is not equal to the corresponding field of `b` deeply.
+func AssertEveryFieldNotEqual[T any](t *testing.T, valA, valB T, opts ...Option) {
+	h := &staticTestHelper{}
+	for _, opt := range opts {
+		opt.apply(h)
+	}
```

Comment:
These codes can be wrapped to a method `staticTestHelper.applyOptions(opts ...Option)` for reuse.